mono/io-layer/wthreads.c

```diff
@@ -217,6 +219,68 @@ GetCurrentThreadId (void)
 	return MONO_NATIVE_THREAD_ID_TO_UINT (id);
 }
 
+static mono_lazy_init_t sleepex_init = MONO_LAZY_INIT_STATUS_NOT_INITIALIZED;
+static mono_mutex_t sleepex_mutex;
+static mono_cond_t sleepex_cond;
+
+static void
+sleepex_initialize (void)
+{
+	mono_mutex_init (&sleepex_mutex);
+	mono_cond_init (&sleepex_cond, NULL);
+}
+
+static void
+sleepex_interrupt (gpointer data)
+{
+	mono_mutex_lock (&sleepex_mutex);
+	mono_cond_broadcast (&sleepex_cond);
+	mono_mutex_unlock (&sleepex_mutex);
+}
+
+static guint32
+sleepex_interruptable (guint32 ms)
```

Comment:
This is the wrong way to handle time wrap around. Ticks wrap around is a lot more common than you'd think.
It happens every 1193 hours, or every 49 days.

The condition to check is not 'now < end' but 'now - start >= ms'.

now - start works fine even in the case of overflow due to how unsigned 2-complement math works.

Let's see how it works with a 1 byte unsigned value. start is 0xFE and we have advanced 1, 2 and 5 (values of now).

0xFE + 1 => 0xFF
0xFE + 2 => 0x00
0xFE + 5 => 0x03

Now let's do the reverse operation (values of now - start):

0xFF - 0xFE => 1
0x00 - 0xFE => 2
0x03 - 0xFE => 5

This ensures that as long as ms is smaller that MAXUINT, we can handle overflow just fine.